torch/csrc/jit/passes/peephole.cpp

```diff
@@ -165,6 +165,51 @@ void PeepholeOptimizeImpl(Block* block, bool addmm_fusion_enabled) {
           u.user->replaceInput(0, node->inputs().at(0));
         }
       }
+    } else if (node->kind() == prim::If) {
+      IfView n(node);
+      // this handles redundant short circuits like "x and True" or "x or False"
+      for (size_t i = 0; i < n.outputs().size(); ++i) {
+        if (n.outputs().at(i)->type() != BoolType::get()) {
+          continue;
+        }
+        bool true_val =
+            constant_as<bool>(n.thenOutputs().at(i)).value_or(false);
+        bool false_val =
+            constant_as<bool>(n.elseOutputs().at(i)).value_or(true);
+        // if an if node's output equals its condition replace output with
+        // condition
+        if (true_val && !false_val) {
+          n.outputs().at(i)->replaceAllUsesWith(n.cond());
+        }
+      }
+    } else if (
+        node->kind() == aten::__is__ || node->kind() == aten::__isnot__) {
+      // if we are comparing a None value with a value that can't be None
+      // replace the output with true if node is __isnot__ or false if node is
+      // __is__
+      AT_ASSERT(node->inputs().size() == 2);
+      for (size_t check_none_index : {0, 1}) {
+        bool input_must_be_none =
+            node->inputs().at(check_none_index)->mustBeNone();
+        Value* other_input = node->inputs().at(1 - check_none_index);
+        bool other_must_not_be_none = !other_input->mustBeNone() &&
+            !other_input->type()->cast<OptionalType>();
```

Comment:
What is the role of `!other_intput->mustBeNone()` here?
I don't think that it is a good idea to use `! mustBeNone` as `cannotBeNone`.
Specifically, ! mustBeNone translates into producing node conditons of [`kind_ != prim::Constant || this->hasAttributes() || ! (output()->type()->cast<OptionalType>() || output()->type() == NoneType::get());`
](https://github.com/pytorch/pytorch/blob/1323c193edf97075bab63759cf0f353b7e7baa2d/torch/csrc/jit/ir.cpp#L773). In my opinion, the check should instead be for NoneType, OptionalType, and AutogradZeroTensorType (which might show up in backwards as None).
Maybe we should define a "canBeNone" helper function in jit_types.h.

As is, this will cause bugs in combination with specializing Optionals to None or AutogradZeroTensor.